Sources/Hummingbird/Middleware/LogRequestMiddleware.swift

```diff
@@ -12,32 +12,111 @@
+        public typealias ArrayLiteralElement = HTTPField.Name
+
+        /// ExpressibleByArrayLiteral requirement
+        public init(arrayLiteral elements: ArrayLiteralElement...) {
+            self = .some(elements)
+        }
+
+        case none
+        case all(except: [HTTPField.Name] = [])
+        case some([HTTPField.Name])
+    }
+
     let logLevel: Logger.Level
-    let includeHeaders: Bool
+    let includeHeaders: HeaderFilter
+    let redactHeaders: [HTTPField.Name]
 
-    public init(_ logLevel: Logger.Level, includeHeaders: Bool = false) {
+    public init(_ logLevel: Logger.Level, includeHeaders: HeaderFilter = .none, redactHeaders: [HTTPField.Name] = []) {
         self.logLevel = logLevel
         self.includeHeaders = includeHeaders
+        // only include headers in the redaction list if we are outputting them
+        self.redactHeaders = switch includeHeaders {
+        case .all(let except):
+            // don't include headers in the except list
+            redactHeaders.filter { header in except.first { $0 == header } == nil }
+        case .some(let included):
+            // only include headers in the included list
+            redactHeaders.filter { header in included.first { $0 == header } != nil }
+        case .none:
+            []
+        }
     }
 
     public func handle(_ request: Request, context: Context, next: (Request, Context) async throws -> Response) async throws -> Response {
-        if self.includeHeaders {
+        switch self.includeHeaders {
+        case .none:
+            context.logger.log(
+                level: self.logLevel,
+                "Request",
+                metadata: [
+                    "hb_uri": .stringConvertible(request.uri),
+                    "hb_method": .string(request.method.rawValue),
+                ]
+            )
+        case .all(let except):
             context.logger.log(
                 level: self.logLevel,
-                "\(request.headers)",
-                metadata: ["hb_uri": .stringConvertible(request.uri), "hb_method": .string(request.method.rawValue)]
+                "Request",
+                metadata: [
+                    "hb_uri": .stringConvertible(request.uri),
+                    "hb_method": .string(request.method.rawValue),
+                    "hb_headers": .string(self.allHeaders(headers: request.headers, except: except)),
+                ]
             )
-        } else {
+        case .some(let filter):
             context.logger.log(
                 level: self.logLevel,
-                "",
-                metadata: ["hb_uri": .stringConvertible(request.uri), "hb_method": .string(request.method.rawValue)]
+                "Request",
+                metadata: [
+                    "hb_uri": .stringConvertible(request.uri),
+                    "hb_method": .string(request.method.rawValue),
+                    "hb_headers": .string(self.filterHeaders(headers: request.headers, filter: filter)),
```

Comment:
Not a fan of formatting the headers instead of passing an e.g. Array or Dictionary as `.stringConvertible`.
We use https://github.com/Brainfinance/StackdriverLogging as our swift-log backend which formats stuff properly as a JSON if it can, which makes them look decent on Google Cloud. By passing a String, a logging backend can't try to format the values anymore.

To be clear i haven't seen too many logging backends try to manually format metadata (my own `DiscordLogger` doesn't either), but it does look useful when done appropriately, like with https://github.com/Brainfinance/StackdriverLogging.

Reply:
Ok makes sense. I can output a dictionary with values from duplicate keys concatenated with a "," inbetween

Reply:
done